cores/esp8266/base64.cpp

```diff
@@ -39,33 +38,20 @@ String base64::encode(const uint8_t * data, size_t length, bool doNewLines) {
     // base64 needs more size then the source data, use cencode.h macros
     size_t size = ((doNewLines ? base64_encode_expected_len(length)
                                : base64_encode_expected_len_nonewlines(length)) + 1);
-    char * buffer = (char *) malloc(size);
-    if(buffer) {
+    String base64("-FAIL-");
+    if (base64.reserve(size)) {
+
         base64_encodestate _state;
-        if(doNewLines)
-        {
+        if(doNewLines) {
             base64_init_encodestate(&_state);
         }
-        else 
-        {
+        else {
             base64_init_encodestate_nonewlines(&_state);
         }
-        int len = base64_encode_block((const char *) &data[0], length, &buffer[0], &_state);
-        len = base64_encode_blockend((buffer + len), &_state);
-
-        String base64 = String(buffer);
-        free(buffer);
-        return base64;
+        int len = base64_encode_block((const char *) &data[0], length, (char*)base64.begin(), &_state);

```

Comment:
Instead of this, can we have base64_encode_block take a String instead and use the `String += <char>` way of setting it.  That keeps base64 out of String's innards, while only requiring the single allocation of the block w/String::reserve.

Reply:
well, thats the question. in the discussion earlier it was said that libb64/ is part of the API and can't be changed, so I don't want to create yet another wrapper around it. 

if we stop considering libb64/ being part of the API then I'd be happy to fold this into base64.cpp and clean it up and make it use String() directly.